builder/chroot/builder.go

```diff
@@ -195,7 +195,7 @@ type Config struct {
 	// prefix `kms_key_id` with `alias/`.
 	RootVolumeKmsKeyId string `mapstructure:"root_volume_kms_key_id" required:"false"`
 	// what architecture to use when registering the final AMI; valid options
-	// are "arm64", "i386", "x86_64", or "x86_64_mac". Defaults to "x86_64".
+	// are "arm64", "arm64_mac", "i386", "x86_64", or "x86_64_mac". Defaults to "x86_64".
```

Comment:
There's no reason why in theory a Linux builder couldn't mount an AMI for another OS and still work on it. That said, I literally just looked for everywhere `x86_64_mac` had been added and included `arm64_mac` in the same places. My particular use case is the ebs-surrogate builder which I'm now successfully using (with this diff) to mount and build a macOS AMI. (It requires installing the experimental https://github.com/linux-apfs/linux-apfs-rw Linux kernel module to do that.)

Reply:
I might be wrong but `chroot` continues using the same kernel, so I'm not certain how chooting into a macOS rootfs would work?

Also (and it may have been forgotten when `x86_64_mac` was added), the `import` post-processor supports an `Architecture` parameter, and I can imagine this might very well support publishing macOS images. The documentation doesn't mention it though, so I can be wrong, please disregard if so.

Reply:
I added `arm64_mac` to be consistent with the existing code, but I don't care either way. I'm using the ebs-surrogate plugin, which again, works with this change to create a macOS AMI from a Linux builder. Without this change that was impossible.

Reply:
I still think one could make the chroot plugin work. You'd need to copy into the chroot all of the Linux bits which would be fine as long as they don't overlap in the filesystem with macOS library/binaries. In any case, I don't see why the chroot plugin should try to prevent it.

Reply:
Background: the reason I'm doing this in the first place is that I need to alter parts of the macOS AMI that are immutable when booted off that AMI. A EC2 Linux instance also boots much faster and costs much less than an EC2 Mac instance to run. To give an idea, packer is able to ssh into a Linux EC2 instance in < 60 seconds. A Mac EC2 instance isn't available for about 10 minutes.

My inspiration for attempting to do this came from a couple places:

- https://github.com/inloco/macami/
- https://jen20.dev/post/building-zfs-root-ubuntu-amis-with-packer/

Reply:
> Subsidiary question: have you tested that change with an ebssurrogate build by any chance? 

Yes, I've been using the plugin modified with this PR for about two weeks now. The only tricky part is getting the `launch_block_device_mappings` and `ami_root_device` correct where I have to pass in information from outside of the packer build. It looks like this:

```
variable "source_ami" {
  # This should match the "Result" TypedDict in stage1-ami-info
  type = object({
    device_name = string
    snapshot_id = string
    source_name = string
    volume_size = number
    volume_type = string
  })
}

source "amazon-ebssurrogate" "ami-stage1" {
  # ... boilerplate elided ...

  # Mount source-ami as /dev/sdf
  launch_block_device_mappings {
    device_name           = "/dev/sdf"
    snapshot_id           = var.source_ami.snapshot_id
    volume_size           = var.source_ami.volume_size
    volume_type           = var.source_ami.volume_type
    delete_on_termination = true
    encrypted             = true
    kms_key_id            = "alias/mac-ec2"
    iops                  = 3000
    throughput            = 125
  }

  # Copy /dev/sdf to a new ami after the instance is terminated.
  ami_root_device {
    source_device_name    = "/dev/sdf"
    device_name           = var.source_ami.device_name
    volume_size           = var.source_ami.volume_size
    volume_type           = var.source_ami.volume_type
    delete_on_termination = true
  }

  # https://github.com/hashicorp/packer-plugin-amazon/pull/487
  ami_architecture = "arm64_mac"
}
```

Then before starting the build I use a Python script to generate `source_ami`. Excerpt:

```python
def sd_mac_source_ami_snapshot_id(macos_version: str = "14.*") -> None:
    source_ami_filter: FilterQuery = {
        "name": [f"amzn-ec2-macos-{macos_version}-*"],
        "architecture": ["arm64_mac"],
        "root-device-type": ["ebs"],
        "virtualization-type": ["hvm"],
    }
    images = ec2.describe_images(
        Owners=["amazon"],
        Filters=make_filter_list(source_ami_filter),
    )["Images"]
    images.sort(key=itemgetter("Name"))

    debug_print("Matching AMIs:")
    debug_print([i["Name"] for i in images])

    image = images[-1]
    debug_print("Chosen AMI:")
    debug_print(image)

    device = image["BlockDeviceMappings"][0]
    source_ami = SourceAMI(
        name=make_image_name(image["Name"]),
        device_name=device["DeviceName"],
        snapshot_id=device["Ebs"]["SnapshotId"],
        source_name=image["Name"],
        volume_size=device["Ebs"]["VolumeSize"],
        volume_type=device["Ebs"]["VolumeType"],
    )
    print(source_ami)
```

I do that outside of packer because the `amazon-ami` data source doesn't provide that information. There's a fork which does but it hasn't been contributed back here:

https://github.com/inloco/packer-plugin-amazon

Reply:
Thanks for the confirmation @jaysoffian !

I'm still having a hard time thinking of how `chroot` would work to prop up a macOS VM, but I get your point that the builder should not preemptively reject that if someone found a way to make it work.

Supporting this for `ebssurrogate` makes sense though, if you're able to build your AMI with this, by all means this makes sense that we support it.

With that in mind, the code looks good to me, only those two support building macOS VMs in theory, so that makes sense to add it to the list of supported archs.

Thanks for sharing the fork also, good to know that those details are missing from our datasource, this is something we should be able to implement eventually.

I'll approve and merge your PR ASAP, I see one of the checks fail on CI because the generated documentation is out-of-sync with the code, could you run `make generate` on the repo, and push the changes?
If you cannot for some reason let me know, and I'll do it for you.

Thanks for the discussion and the feature!

Reply:
> I'll approve and merge your PR ASAP, I see one of the checks fail on CI because the generated documentation is out-of-sync with the code, could you run make generate on the repo, and push the changes?

Done!

Thank for you for the review.